### PR TITLE
Make OhmmsVector safe to non-trivial types.

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Make `Vector<std::vector>` work.
It didn't work due to uninitialized Vector element (type std::vector) after allocating memory.
The fix is using placement new to make sure non-trival types are constructed.

## What type(s) of changes does this code introduce?
- Bugfix

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

**Comment by ye-luo** [comment]
Test this please

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
In prinicipal this is a great idea.  I just want time to do a decent review and check DualAllocator behavior with it.

File: src/Containers/OhmmsPETE/OhmmsVector.h [hunk 1]
```diff
@@ -301,6 +290,34 @@ class Vector
...
+    if constexpr (qmc_allocator_traits<Alloc>::is_host_accessible)
+    {
+      if constexpr (std::is_trivial<T>::value)
+        std::copy_n(from, n_elements, to);
+      else
+        for (size_t i = 0; i < n_elements; i++)
+          new (to + i) Type_t(*(from + i));
+    }
+  }
+
```
Comment (PDoakORNL) [inline src/Containers/OhmmsPETE/OhmmsVector.h:314]: There was once the assumption you could marshal OhmmsContainers over MPI but just copying the bytes in X. Or attaching X to a part of a big buffer and then sending that buffer.  You just can't do that even with this more correct code.  I think the answer to that is just code review to keep such schemes from returning.

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
I think the unit test  just needs a little expansion since it seems to be illustrating the OhmmsStyle resize behavior. At the same time it should document not to depend on this I think. So the unit test fails if the behavior is changed accidentally and the unit test can be changed if the resize behavior is updated. 

An integration test with Dual Allocators in a future PR will be important.

File: src/Containers/OhmmsPETE/OhmmsVector.h [hunk 2]
```diff
@@ -187,16 +171,19 @@ class Vector
 
     if (n > nAllocated)
     {
+      if (nLocal)
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/Containers/OhmmsPETE/OhmmsVector.h:174]: I think we require Vector to not preserving data when resize happens. Even just resize to the current size, developers should not rely on the old value. This point needs to be clear in the source code comments.

File: src/Containers/OhmmsPETE/OhmmsVector.h (hunk 1 above)
Comment (ye-luo) [reply src/Containers/OhmmsPETE/OhmmsVector.h:314]: I think you can still do that with MPI as long as you guarantee type safety. I don't get you mean here.

File: src/Containers/OhmmsPETE/tests/test_Vector.cpp [hunk 3]
```diff
@@ -98,4 +99,27 @@ TEST_CASE("VectorViewer", "[OhmmsPETE]")
   // TODO: add optional bounds checking to accesses via operator[]
 }
 
+TEST_CASE("NestedContainers", "[OhmmsPETE]")
+{
+  Vector<std::vector<int>> vec_of_vecs(2);
+  vec_of_vecs[0].push_back(123);
+  vec_of_vecs.resize(5);
+  vec_of_vecs[0].push_back(123);
+  vec_of_vecs.resize(0);
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/Containers/OhmmsPETE/tests/test_Vector.cpp:108]: Indeed. You caught a bug in my test. I said users should not rely on previous data but push_back by nature relies on the previous data. So I need explicit clear before using it.

File: src/Containers/OhmmsPETE/tests/test_Vector.cpp [hunk 4]
```diff
@@ -98,4 +99,27 @@ TEST_CASE("VectorViewer", "[OhmmsPETE]")
...
+  vec_of_vecs.resize(3);
+  vec_of_vecs[0].push_back(123);
+  CHECK(vec_of_vecs[0].back() == 123);
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/Containers/OhmmsPETE/tests/test_Vector.cpp:111]: If it is destroyed, you cannot access `vec_of_vecs[0]` thus cannot perform the check you requested.

File: src/Containers/OhmmsPETE/tests/test_Vector.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/Containers/OhmmsPETE/tests/test_Vector.cpp:108]: >  But I think I expect that to fail here.
In the current implementation, it fails.

**Review by PDoakORNL — APPROVED** [review]

File: src/Containers/OhmmsPETE/OhmmsVector.h (hunk 1 above)
Comment (ye-luo) [reply src/Containers/OhmmsPETE/OhmmsVector.h:315]: Correct. Only non pointer value can be sent via MPI,

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**